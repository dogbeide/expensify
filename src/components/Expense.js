import { Link } from 'react-router-dom'
import React from 'react'

import { removeExpense } from '../actions/expenses'


const Expense = ({description, amount, createdAt, id}) => (
    <div>
        <Link to={`/edit/${id}`} >
            <h3>{description}</h3>
        </Link>
        <p>{amount} cents </p>
        <p>Created at: {createdAt}</p>
        <hr/>
    </div>
)

export default Expense