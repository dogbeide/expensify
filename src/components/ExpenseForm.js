import moment from 'moment'
import React from 'react'
import { SingleDatePicker } from 'react-dates'
import 'react-dates/initialize'
import expensesReducer from '../reducers/expenses';


export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props)
        this.state = {
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            description: props.expense ? props.expense.description : '',
            error: '',
            calendarFocused: false,
            note: props.expense ? props.expense.note : '',
        }
    }


    onDescriptionChange = (e) => {
        const description = e.target.value
        this.setState(() => ({ description }))
    }

    onNoteChange = (e) => {
        // Directly accessing synthetic events in callback gives warning
        // must assign or persist first

        // const note = e.target.value
        e.persist()
        this.setState(() => ({ note: e.target.value }))
    }

    onAmountChange = (e) => {
        const amount = e.target.value
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }))
        }
    }

    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({ createdAt }))
        }
    }

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }))
    }

    onSubmit = (e) => {
        e.preventDefault()

        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: 'Please add description and amount' }))
        } else {
            this.setState(() => ({ error: ''}))
            this.props.onSubmit({
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                description: this.state.description,
                note: this.state.note
            })
        }
    }

    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input
                        autoFocus
                        onChange={this.onDescriptionChange}
                        placeholder="Description"
                        type="text"
                        value={this.state.description}
                    />
                    <input
                        onChange={this.onAmountChange}
                        placeholder="Amount"
                        // type="number" // too many decimals
                        type="text"
                        value={this.state.amount}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        focused={this.state.calendarFocused}
                        isOutsideRange={() => false}
                        numberOfMonths={1}
                        onDateChange={this.onDateChange}
                        onFocusChange={this.onFocusChange}
                    />
                    <textarea
                        id=""
                        name=""
                        onChange={this.onNoteChange}
                        placeholder="Additional notes (optional)"
                        value={this.state.note}
                    >
                    </textarea>
                    <button>Add Expense</button>
                </form>
            </div>
        )
    }
}