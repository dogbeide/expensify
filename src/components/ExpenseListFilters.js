import React from "react";
import { connect } from 'react-redux'
import { DateRangePicker } from 'react-dates'

import {
    setEndDate,
    setStartDate,
    setTextFilter,
    sortByAmount,
    sortByDate
} from '../actions/filters'


export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    }

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }))
    }

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value)
    }

    onSortByChange = (e) => {
        if (e.target.value === 'amount') {
            this.props.sortByAmount()
        } else if (e.target.value === 'date') {
            this.props.sortByDate()
        }
    }

    render() {
        return (
            <div>
                <input 
                    type="text"
                    defaultValue={this.props.filters.text}
                    onChange={this.onTextChange}
                />
                <select
                    value={this.props.filters.sortBy}
                    onChange={this.onSortByChange}
                >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                    endDate={this.props.filters.endDate}
                    focusedInput={this.state.calendarFocused}
                    isOutsideRange={() => false}
                    numberOfMonths={1}
                    onDatesChange={this.onDatesChange}
                    onFocusChange={this.onFocusChange}
                    showClearDates={true}
                    startDate={this.props.filters.startDate}
                    startDateId="startDate"
                    endDateId="endDate"
                />
            </div>
        )
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
}

const mapDispatchToProps = (dispatch) => ({
    setEndDate: (date) => dispatch(setEndDate(date)),
    setStartDate: (date) => dispatch(setStartDate(date)),
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByAmount: () => dispatch(sortByAmount()),
    sortByDate: () => dispatch(sortByDate())
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)