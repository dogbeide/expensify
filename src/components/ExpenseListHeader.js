import React from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import numeral from 'numeral'

import selectExpenses from '../selectors/expenses'
import getExpensesTotal from '../selectors/total'


export const ExpenseListHeader = ({ count, total }) => (
    <div className="page-header">
        <div className="content-container">
            <h1 className="page-header__title">
                Showing <span>{count}</span>
                {` expense${count !== 1 ? 's' : ''} `}
                totaling: <span>{numeral(total/100).format('$0,0.00')}</span>
            </h1>
            <div className="page-header__actions">
                <Link className="button" to="/create">Add Expense</Link>
            </div>
        </div>
    </div>
)

const mapStateToProps = (state) => {
    const selected = selectExpenses(state.expenses, state.filters) 
    return {
        count: selected.length,
        total: getExpensesTotal(selected)
    }
}

export default connect(mapStateToProps)(ExpenseListHeader)
