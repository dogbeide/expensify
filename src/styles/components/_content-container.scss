.content-container {
    margin: 0 auto;
    padding: 0 $size-md;
    max-width: 80rem;
}
