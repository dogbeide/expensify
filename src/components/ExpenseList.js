import React from 'react'
import { connect } from 'react-redux'

import Expense from './Expense'
import ExpenseListHeader from './ExpenseListHeader'
import getExpensesTotal from '../selectors/total'
import selectExpenses from '../selectors/expenses'


export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (
                <p>No expenses</p>
            ) : (
                props.expenses.map((expense) => (
                    <Expense key={expense.id} {...expense} />
                ))
            )
        }
    </div>
)

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
}

export default connect(mapStateToProps)(ExpenseList)