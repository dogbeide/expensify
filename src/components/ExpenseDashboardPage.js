import React from 'react'

import ExpenseList from './ExpenseList'
import ExpenseListFilters from './ExpenseListFilters'
import ExpenseListHeader from './ExpenseListHeader'

const ExpenseDashboardPage = () => (
    <div>
        <ExpenseListFilters />
        <ExpenseListHeader />
        <ExpenseList />
    </div>
)

export default ExpenseDashboardPage