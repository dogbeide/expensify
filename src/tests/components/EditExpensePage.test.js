import React from 'react'
import { shallow } from 'enzyme';

import { EditExpensePage } from '../../components/EditExpensePage'
import expenses from '../fixtures/expenses'


let wrapper, editExpense, removeExpense, history

beforeEach(() => {
    history = { push: jest.fn() }
    editExpense = jest.fn()
    removeExpense = jest.fn()
    wrapper = shallow(
        <EditExpensePage
            expense={expenses[0]}
            editExpense={editExpense}
            removeExpense={removeExpense}
            history={history}
        />
    )
})

test('should render EditExpensePage', () => {
    expect(wrapper).toMatchSnapshot()
})

test('should handle editExpense', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1])
    expect(editExpense).toHaveBeenLastCalledWith(
        expenses[0].id,
        expenses[1]
    )
    expect(history.push).toHaveBeenLastCalledWith('/')
})

test('should handle removeExpense', () => {
    wrapper.find('button').simulate('click')
    expect(removeExpense).toHaveBeenLastCalledWith({
        id: expenses[0].id
    })
    expect(history.push).toHaveBeenLastCalledWith('/')
})