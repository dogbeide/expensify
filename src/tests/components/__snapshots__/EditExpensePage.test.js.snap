// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 666,
        "createdAt": 0,
        "description": "Yums",
        "id": "1",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
